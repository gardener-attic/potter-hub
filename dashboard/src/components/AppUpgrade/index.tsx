import AppUpgrade from "./AppUpgrade";

export default AppUpgrade;
