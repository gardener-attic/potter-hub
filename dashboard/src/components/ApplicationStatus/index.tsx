import ApplicationStatus from "./ApplicationStatus";

export default ApplicationStatus;
