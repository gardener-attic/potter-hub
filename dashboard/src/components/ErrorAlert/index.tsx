import ErrorAlertHeader from "./ErrorAlertHeader";
import ErrorSelector from "./ErrorSelector";
import MessageAlert from "./MessageAlert";
import NotFoundErrorAlert from "./NotFoundErrorAlert";
import PermissionsErrorAlert from "./PermissionsErrorAlert";
import ServiceBrokersNotFoundAlert from "./ServiceBrokersNotFoundAlert";
import ServiceCatalogNotInstalledAlert from "./ServiceCatalogNotInstalledAlert";
import UnexpectedErrorAlert from "./UnexpectedErrorAlert";

export {
  ErrorAlertHeader,
  ErrorSelector,
  MessageAlert,
  NotFoundErrorAlert,
  PermissionsErrorAlert,
  ServiceBrokersNotFoundAlert,
  ServiceCatalogNotInstalledAlert,
  UnexpectedErrorAlert,
};
