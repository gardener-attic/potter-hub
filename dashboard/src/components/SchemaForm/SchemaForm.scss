.SchemaForm > .form-group > fieldset {
  padding: 0;
  border: none;
}
