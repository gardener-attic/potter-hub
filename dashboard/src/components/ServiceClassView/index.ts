import ServiceClassView from "./ServiceClassView";

export default ServiceClassView;
