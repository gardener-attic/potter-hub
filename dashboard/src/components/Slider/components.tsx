// CODE FROM: https://sghall.github.io/react-compound-slider/#/slider-demos/horizontal
import * as React from "react";
import { GetHandleProps, GetTrackProps, SliderItem } from "react-compound-slider";

// *******************************************************
// HANDLE COMPONENT
// *******************************************************
interface IHandleProps {
  domain: number[];
  handle: SliderItem;
  getHandleProps: GetHandleProps;
}

export const Handle: React.SFC<IHandleProps> = ({
  domain: [min, max],
  handle: { id, value, percent },
  getHandleProps,
}) => (
  <div
    role="slider"
    aria-valuemin={min}
    aria-valuemax={max}
    aria-valuenow={value}
    style={{
      left: `${percent}%`,
      position: "absolute",
      marginLeft: "-11px",
      marginTop: "-6px",
      zIndex: 2,
      width: 24,
      height: 24,
      cursor: "pointer",
      borderRadius: "50%",
      boxShadow: "1px 1px 1px 1px rgba(0, 0, 0, 0.2)",
      backgroundColor: "#34568f",
    }}
    {...getHandleProps(id)}
  />
);

// *******************************************************
// TRACK COMPONENT
// *******************************************************
interface ITrackProps {
  source: SliderItem;
  target: SliderItem;
  getTrackProps: GetTrackProps;
}

export const Track: React.SFC<ITrackProps> = ({ source, target, getTrackProps }) => (
  <div
    style={{
      position: "absolute",
      height: 14,
      zIndex: 1,
      backgroundColor: "#7aa0c4",
      borderRadius: 7,
      cursor: "pointer",
      left: `${source.percent}%`,
      width: `${target.percent - source.percent}%`,
    }}
    {...getTrackProps()}
  />
);
