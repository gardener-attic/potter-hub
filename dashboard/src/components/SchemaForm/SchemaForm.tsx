import { JSONSchema6 } from "json-schema";
import * as React from "react";
import Form, { ISubmitEvent } from "react-jsonschema-form";

import ArrayFieldTemplate from "./ArrayFieldTemplate";
import CustomObjectField from "./CustomObjectField";
import FieldTemplate from "./FieldTemplate";
import "./SchemaForm.css";

interface ISchemaFormProps {
  schema: JSONSchema6;
  onSubmit?: (result: ISubmitEvent<any>) => void;
}

class SchemaForm extends React.Component<ISchemaFormProps> {
  public render() {
    return (
      <Form
        className="SchemaForm"
        schema={this.props.schema}
        onSubmit={this.props.onSubmit}
        autocomplete="off"
        FieldTemplate={FieldTemplate}
        ArrayFieldTemplate={ArrayFieldTemplate}
        fields={{ ObjectField: CustomObjectField }}
        noValidate={true}
      >
        {this.props.children}
      </Form>
    );
  }
}

export default SchemaForm;
