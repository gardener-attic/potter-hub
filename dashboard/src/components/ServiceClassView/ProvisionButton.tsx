import { RouterAction } from "connected-react-router";
import * as React from "react";
import Modal from "react-modal";

import { IClusterServiceClass } from "../../shared/ClusterServiceClass";
import { IServicePlan } from "../../shared/ServiceCatalog";
import { IRBACRole } from "../../shared/types";
import { ErrorSelector } from "../ErrorAlert";
import SchemaForm from "../SchemaForm";

import { JSONSchema6 } from "json-schema";
import { ISubmitEvent } from "react-jsonschema-form";

interface IProvisionButtonProps {
  namespace: string;
  error?: Error;
  selectedClass: IClusterServiceClass;
  selectedPlan: IServicePlan;
  provision: (
    releaseName: string,
    namespace: string,
    className: string,
    planName: string,
    parameters: {},
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
}

interface IProvisionButtonState {
  isProvisioning: boolean;
  modalIsOpen: boolean;
  name: string;
  displayNameForm: boolean;
}

const RequiredRBACRoles: IRBACRole[] = [
  {
    apiGroup: "servicecatalog.k8s.io",
    resource: "serviceinstances",
    verbs: ["create"],
  },
];

class ProvisionButton extends React.Component<IProvisionButtonProps, IProvisionButtonState> {
  public state: IProvisionButtonState = {
    displayNameForm: true,
    isProvisioning: false,
    modalIsOpen: false,
    name: "",
  };

  public render() {
    const { selectedClass, selectedPlan, error, namespace } = this.props;
    let schema = selectedPlan.spec.instanceCreateParameterSchema;
    if (!schema) {
      schema = {
        properties: {
          kubeappsRawParameters: {
            title: "Parameters",
            type: "object",
          },
        },
        type: "object",
      };
    }

    return (
      <div className="ProvisionButton">
        {this.state.isProvisioning && <div>Provisioning...</div>}
        <button
          className="button button-primary button-small"
          onClick={this.openModel}
          disabled={this.state.isProvisioning}
        >
          Provision
        </button>
        <Modal
          isOpen={this.state.modalIsOpen}
          onRequestClose={this.closeModal}
          contentLabel="Modal"
          className={this.state.displayNameForm ? "centered-modal" : ""}
        >
          {error && (
            <ErrorSelector
              error={error}
              resource={selectedClass.spec.externalName}
              action="provision"
              defaultRequiredRBACRoles={{ provision: RequiredRBACRoles }}
              namespace={namespace}
            />
          )}
          {this.state.displayNameForm ? (
            <SchemaForm schema={this.nameSchema()} onSubmit={this.handleNameChange}>
              <div>
                <button className="button button-primary" type="submit">
                  Continue
                </button>
                <button className="button" onClick={this.closeModal}>
                  Cancel
                </button>
              </div>
            </SchemaForm>
          ) : (
            <SchemaForm schema={schema} onSubmit={this.handleProvision}>
              <div>
                <button
                  className="button button-primary"
                  type="submit"
                  disabled={this.state.isProvisioning}
                >
                  Submit
                </button>
                <button
                  className="button"
                  onClick={this.handleBackButton}
                  disabled={this.state.isProvisioning}
                >
                  Back
                </button>
              </div>
            </SchemaForm>
          )}
        </Modal>
      </div>
    );
  }

  public openModel = () => {
    this.setState({
      modalIsOpen: true,
    });
  };

  public closeModal = () => {
    this.setState({
      modalIsOpen: false,
    });
  };

  public handleBackButton = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    this.setState({ displayNameForm: true });
  };

  public handleNameChange = ({ formData }: ISubmitEvent<{ Name: string }>) => {
    this.setState({ name: formData.Name, displayNameForm: false });
  };

  public handleProvision = async ({
    formData,
  }: ISubmitEvent<{ name: string; kubeappsRawParameters: {} }>) => {
    const { namespace, provision, push, selectedClass, selectedPlan } = this.props;
    const { name } = this.state;
    this.setState({ isProvisioning: true });

    const { kubeappsRawParameters, ...rest } = formData;
    if (selectedClass && selectedPlan) {
      const provisioned = await provision(
        name,
        namespace,
        selectedClass.spec.externalName,
        selectedPlan.spec.externalName,
        kubeappsRawParameters || rest,
      );
      if (provisioned) {
        push(
          `/services/brokers/${selectedClass.spec.clusterServiceBrokerName}/instances/ns/${namespace}/${name}`,
        );
      } else {
        this.setState({ isProvisioning: false });
      }
    }
  };

  private nameSchema(): JSONSchema6 {
    return {
      properties: {
        Name: {
          default: this.state.name,
          description: "Name for ServiceInstance",
          type: "string",
        },
      },
      required: ["Name"],
      type: "object",
    };
  }
}

export default ProvisionButton;
