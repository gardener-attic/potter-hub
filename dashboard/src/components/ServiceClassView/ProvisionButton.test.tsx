import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import { JSONSchema6 } from "json-schema";
import { ErrorSelector } from "../../components/ErrorAlert";
import SchemaForm from "../../components/SchemaForm";
import { IClusterServiceClass } from "../../shared/ClusterServiceClass";
import { IServicePlan } from "../../shared/ServiceCatalog";
import ProvisionButton from "./ProvisionButton";

const defaultName = "my-class";
const defaultNS = "default";
const selectedClass = {
  metadata: {
    name: defaultName,
    uid: `class-${defaultName}`,
  },
  spec: {
    bindable: true,
    externalName: defaultName,
    description: "this is a class",
    externalMetadata: {
      imageUrl: "img.png",
    },
  },
} as IClusterServiceClass;
let selectedPlan: any = {};
let defaultProps = {
  selectedClass,
  selectedPlan,
  provision: jest.fn(),
  push: jest.fn(),
  namespace: defaultNS,
};

beforeEach(() => {
  selectedPlan = {
    metadata: {
      name: defaultName,
    },
    spec: {
      externalName: defaultName,
      externalID: `plan-${defaultName}`,
      description: "this is a plan",
      clusterServiceClassRef: {
        name: defaultName,
      },
      free: true,
    },
  } as IServicePlan;
  defaultProps = {
    selectedClass,
    selectedPlan,
    provision: jest.fn(),
    push: jest.fn(),
    namespace: defaultNS,
  };
});

it("shows a provisioning message", () => {
  const wrapper = shallow(<ProvisionButton {...defaultProps} />);
  wrapper.setState({ isProvisioning: true });
  expect(wrapper.text()).toContain("Provisioning...");
});

context("when the modal is open", () => {
  it("shows an error if is present", () => {
    const wrapper = shallow(<ProvisionButton {...defaultProps} error={new Error()} />);
    wrapper.setState({ modalIsOpen: true });

    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper).toMatchSnapshot();
  });

  it("should display the name form", () => {
    const wrapper = shallow(<ProvisionButton {...defaultProps} />);
    wrapper.setState({ modalIsOpen: true, displayNameForm: true });

    const form = wrapper.find(SchemaForm);
    expect(form).toExist();
    expect((form.prop("schema") as any).properties).toHaveProperty("Name");
  });

  it("should display the default schema if it's not in the plan spec", () => {
    selectedPlan.spec.instanceCreateParameterSchema = null;
    const wrapper = shallow(<ProvisionButton {...defaultProps} />);
    wrapper.setState({ modalIsOpen: true, displayNameForm: false });

    const form = wrapper.find(SchemaForm);
    expect(form).toExist();
    const schema = form.prop("schema") as any;
    expect(schema).toMatchObject({
      properties: {
        kubeappsRawParameters: {
          title: "Parameters",
          type: "object",
        },
      },
      type: "object",
    });
  });

  it("should display the spec schema", () => {
    const newSchema = {
      properties: {
        kubeappsRawParameters: {
          title: "Foo",
          type: "string",
        },
      },
      type: "object",
    } as JSONSchema6;
    selectedPlan.spec.instanceCreateParameterSchema = newSchema;
    const wrapper = shallow(<ProvisionButton {...defaultProps} />);
    wrapper.setState({ modalIsOpen: true, displayNameForm: false });

    const form = wrapper.find(SchemaForm);
    expect(form).toExist();
    const schema = form.prop("schema") as any;
    expect(schema).toEqual(newSchema);
  });
});
