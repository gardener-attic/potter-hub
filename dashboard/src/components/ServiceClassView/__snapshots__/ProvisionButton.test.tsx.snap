// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the modal is open shows an error if is present 1`] = `
<div
  className="ProvisionButton"
>
  <button
    className="button button-primary button-small"
    disabled={false}
    onClick={[Function]}
  >
    Provision
  </button>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="centered-modal"
    closeTimeoutMS={0}
    contentLabel="Modal"
    isOpen={true}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <ErrorSelector
      action="provision"
      defaultRequiredRBACRoles={
        Object {
          "provision": Array [
            Object {
              "apiGroup": "servicecatalog.k8s.io",
              "resource": "serviceinstances",
              "verbs": Array [
                "create",
              ],
            },
          ],
        }
      }
      error={[Error]}
      namespace="default"
      resource="my-class"
    />
    <SchemaForm
      onSubmit={[Function]}
      schema={
        Object {
          "properties": Object {
            "Name": Object {
              "default": "",
              "description": "Name for ServiceInstance",
              "type": "string",
            },
          },
          "required": Array [
            "Name",
          ],
          "type": "object",
        }
      }
    >
      <div>
        <button
          className="button button-primary"
          type="submit"
        >
          Continue
        </button>
        <button
          className="button"
          onClick={[Function]}
        >
          Cancel
        </button>
      </div>
    </SchemaForm>
  </Modal>
</div>
`;
